rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java

```diff
@@ -235,6 +245,115 @@ public InterruptibleFuture<ITuple> getRename(ITree module, Position cursor, Set<
+
+        var writer = VF.setWriter();
+        for (var rename : renames) {
+            ISourceLocation currentLoc = sourceLocationFromUri(rename.getOldUri());
+            ISourceLocation newLoc = sourceLocationFromUri(rename.getNewUri());
+
+            ISourceLocation currentWsFolder = findContainingWorkspaceFolder(currentLoc, sortedWorkspaceFolders)
+                .orElseThrow(() -> new RuntimeException(String.format("Cannot automatically change uses of %s, since it is outside the current workspace.", currentLoc)));
+
+            ISourceLocation newWsFolder = findContainingWorkspaceFolder(newLoc, sortedWorkspaceFolders)
+                .orElseThrow(() -> new RuntimeException(String.format("Cannot automatically change uses of %s, since that it is outside the current workspace.", newLoc)));
+
+            if (!currentWsFolder.equals(newWsFolder)) {
+                String commonProjPrefix = StringUtils.getCommonPrefix(currentWsFolder.toString(), newWsFolder.toString());
+                String currentProject = StringUtils.removeStart(currentWsFolder.toString(), commonProjPrefix);
+                String newProject = StringUtils.removeStart(newWsFolder.toString(), commonProjPrefix);
+
+                throw new RuntimeException(String.format("Cannot automatically change uses of %s, since moving files between projects (from %s to %s) is not supported", currentLoc, currentProject, newProject));
+            }
+
+            PathConfig pcfg = getPathConfig.apply(currentWsFolder);
+            try {
+                IString currentName = VF.string(pcfg.getModuleName(currentLoc));
+                IString newName = VF.string(pcfg.getModuleName(newLoc));
+
+                writer.insert(VF.tuple(currentName, newName, addResources(pcfg)));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        return writer.done();
+    }
+
+    private static String readFile(ISourceLocation loc, String charset) throws IOException {
+        URIResolverRegistry reg = URIResolverRegistry.getInstance();
+        try (Reader reader = reg.getCharacterReader(loc, charset)) {
+            StringBuilder res = new StringBuilder();
+            char[] chunk = new char[8192]; // from Prelude.java
+            int read;
+            while ((read = reader.read(chunk, 0, chunk.length)) != -1) {
+                res.append(chunk, 0, read);
+            }
+            return res.toString();
+        }
+    }
+
+    private CompletableFuture<ITree> getEditorTreeOrParse(ISourceLocation loc, Map<ISourceLocation, TextDocumentState> documents) throws IOException {
+        if(documents.containsKey(loc)) {
+            var file = documents.get(loc);
+            return file.getCurrentTreeAsync()
+                .thenApply(Versioned::get)
+                .handle((t, r) -> (t == null ? file.getMostRecentTree().get() : t))
+                ;
+        }
+
+        var input = readFile(loc, "UTF-8");
```

Comment:
we don't know if the file is encoded in utf8. and always better to use `StandardCharsets.UTF8` instead of a string literal.

also, we should read the file inside the `CompletableFuture`, not outside of it.